### DISCUSSION TITLE
Separate concretization of build dependencies

### DISCUSSION THREAD

**Discussion by alalazo** [spack/spack#21564] [Ideas]
Currently Spack permits at most a single node stemming from any given package in a concretized DAG, independently of the dependency types connecting the nodes. This ensures consistency in the DAG, but may be overly restrictive. For instance, we can't currently concretize a DAG where one node requires Python 2.X to build and another requires Python 3.X to run. 

Below I'll outline a proposal for a generalization of the current concretization algorithm which is:
- Based on the iterative construction of the final DAG  
- Reusing the current solve procedure (with its unification constraints) to construct **parts** of the DAG
- Adding new API calls that complement the subscript spec API to deal with multiple nodes stemming from the same package

The goal is to permit the separate concretization of build dependencies for any root spec.

### Iterative concretization algorithm
In the following pseudo-code `solve` is the current solution algorithm, the `types=` argument are the dependency types accounted for that solve:
```python
# The spec we want to deploy obeys the "at most one node per package" unification constraint
# but is constructed accounting only for link and run type dependencies. 
deploy_spec = solve(root_spec, types=('link', 'run'))

for each node in deploy_spec:
  # Get a list of build requirements for this node
  build_requirements_by_package = build_requirements(node)
  
  # Concretize build specs maximizing the reuse of:
  # 1. Specs already in deploy_spec
  # 2. Specs already in build_specs computed in previous steps
  for each abstract_build_spec in build_requirements_by_package:
      build_spec = solve(abstract_build_spec, types=('build', 'link', 'run'))
      
      # Add build spec to the node and construct the DAG iteratively
      node.add_build_dependency(build_spec)
```

The overall idea is to construct first the spec that will be deployed, and iterate afterwards on each node to complement it with the build dependencies that are needed to build it from sources. The iterative construction should be done in a way to maximise the reuse of specs that have been already computed.

### Discussion on the iterative construction of a DAG

One consideration on this approach is that it simplifies modelling cross-compilation **by dividing a single solve across architectural boundaries into multiple solves that are homogeneous on the "target" architecture**:
1. The spec to be deployed targets the architecture of the machine where is supposed to run
2. The build dependencies target the architecture of the machine where the software is built

Each of the solve is done in a way that minimizes changes to the current logic program and prevents the search space for a solution to explode (as compared to a coupled solution allowing for multiple nodes from the same package). 

Note that "build" dependencies cannot influence anymore the presence or absence of "link" / "run" dependencies, unless we want to enclose the iteration above in a self-consistent loop. While in principle this may be a limitation, in practice it's difficult to find a package where this feedback from "build" dependencies to "link" / "run" dependencies is used. For instance, there's no package - as far as I know - that deploys differently if it is built with `cmake@X.Y:` with respect to `cmake@A.B`:
```python
# CMake is a build dependency
depends_on('cmake', type='build')

# Theoretically this is possible, but not used in practice
depends_on('foo@X.Y', type='link', when='^cmake@A.B:')
```
Also, in most case a condition like the one above can be reformulated as a decision over an intermediate condition that introduces an additional requirement on a build dependency:
```python
depends_on('foo@X.Y', type='link', when=another_condition)
depends_on('cmake@A.B:', type='build', when=another_condition)
```

### Changes required to the Spec API

The "single node per package" constraint is widely depended upon in the current spec API due to the subscripting syntax of specs. Each spec can in fact be used similarly to a dictionary to retrieve other nodes or sub-DAGs, using package names as keys:
```python
python_spec = root_spec['python']
```
Relaxing this constraint to permit multiple nodes from the same package requires to extend the current semantics. We can maintain backward compatibility by ordering the multiple nodes by dependency types so that `link` < `run` < `build` < `test` and then sorting the nodes. This ensures that `link` and `run` specs are returned by the subscript semantics in case there's a split `build` dependency. 

Other API calls can then be added to `Spec` to return selected nodes or edges in the DAG. For instance:
```python
dspecs = spec.dependencies(deptype='build')
```
can be used to get all the build dependencies from a given node. #21683 provides an implementation of these API calls that solves #11983 based on a new data structure to store edges that are connecting a node to dependents or dependencies.

### Default dependency types

Currently the default dependency types are:

https://github.com/spack/spack/blob/3c874e22b591cce8b7d14674f2c1102c29befd3d/lib/spack/spack/dependency.py#L16-L17

Given the semantics of dependency types, it would be probably good to turn it to "link" only (the dependency is needed at compile and/or link and/or load time). Rationale is that not many dependencies act both as a build tool ("build" dependencies) and a library ("link" dependency) at the same time. The default value is possibly vestigial from the initial interpretation of "build" dependency as something needed at build time and the fact that C or C++ header files enter within this definition.

### Related PRs and Issues

- First issue in this sense: see #839
- Tentative implementation done on the original concretizer: see #4595 and #2548
- Issue with `Spec._dependents` mapping a package name to a single dependency spec: see #11983
- Issue related to build dependencies entering the DAG hash under certain conditions: see #21491
- Property of the current dependency model and future roadmap: see https://github.com/spack/build-si-modeling/blob/master/outline.md

### PRs addressing this feature
- [x] #21683 Permits to represent multiple dependency / dependent nodes stemming from the same package

**Comment by cessenat** [comment]
This is very interesting. Recently, flex preferred version has been set to 2.6.3, which is used for mesa build. Meanwhile, Scotch does not build with 2.6.3, but is OK with 2.6.4. So,when compiling an app with both mesa and scotch one gets a concretization error. Your idea would make it work without having to force a specific version of flex for both independent applications, scotch and mesa.

**Reply by alalazo** [reply]
Yes, that's one of use cases that this extension aims at solving.

**Reply by haampie** [reply]
I don't think this should lead to a concretization error, flex 2.6.3 is preferred cause it doesn't need patches, but that  doesn't mean you can't use 2.6.4

**Reply by alalazo** [reply]
I didn't check this particular use case, but the one in which the same build dependency is needed in conflicting version ranges by two nodes in the DAG is covered (e.g. one node requires `cmake@:3.1.0 another node requires `cmake@3.12.0:`). I think a real use case of this kind was reported on Slack and involved the Qt stack.

**Reply by haampie** [reply]
Yeah, that's a very nice feature

**Reply by cessenat** [reply]
Yes that is right, one can use 2.6.4 but is has to be enforced:
`spack spec gmsh+fltk+mmg~med~cgns~gmp ^flex@2.6.4`
(fltk depends on mesa, med depends on scotch).

**Reply by haampie** [reply]
@cessenat you might want to use the new concretizer, it should automatically resolve to the latest version of flex because scotch can't depend on 2.6.3; set `concretizer: clingo` in `spack edit config`

**Reply by cessenat** [reply]
Thank you @haampie ; Unfortunately, I get stuck - sorry here is probably not the right place to mention that - with the message:
```
$ spack spec gmsh+fltk+mmg~med~cgns~gmp~mpi
Input spec
--------------------------------
gmsh~cgns+fltk~gmp~med+mmg~mpi

Concretized
--------------------------------
==> Warning: Missing a source id for python@3.6
==> Installing libiconv-1.16-jearpk4xci4zc7dkrza4fufaqfkq7rfl
==> Fetching https://cache.e4s.io/build_cache/linux-ubuntu18.04-x86_64/gcc-7.5.0/libiconv-1.16/linux-ubuntu18.04-x86_64-gcc-7.5.0-libiconv-1.16-jearpk4xci4zc7dkrza4fufaqfkq7rfl.spack
###################################################################################################### 100.0%
==> Extracting libiconv-1.16-jearpk4xci4zc7dkrza4fufaqfkq7rfl from binary cache
gpgconf: socketdir is '/run/user/501/gnupg'
gpg: Signature made Tue May 11 04:01:43 2021 CEST
gpg:                using RSA key 7D344E2992071B0AAAE1EDB0E68DE2A80314303D
gpg: Good signature from "prl" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 7D34 4E29 9207 1B0A AAE1  EDB0 E68D E2A8 0314 303D
==> Warning: Missing a source id for python@3.6
==> Installing libiconv-1.16-jearpk4xci4zc7dkrza4fufaqfkq7rfl
==> Extracting libiconv-1.16-jearpk4xci4zc7dkrza4fufaqfkq7rfl from binary cache
gpg: Signature made Tue May 11 04:01:43 2021 CEST
gpg:                using RSA key 7D344E2992071B0AAAE1EDB0E68DE2A80314303D
gpg: Good signature from "prl" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 7D34 4E29 9207 1B0A AAE1  EDB0 E68D E2A8 0314 303D
==> Warning: Missing a source id for python@3.6
==> Installing libiconv-1.16-jearpk4xci4zc7dkrza4fufaqfkq7rfl
```
and this goes on again and again.
```
$ spack debug report
* **Spack:** 0.16.1-2696-f1c849ca0f
* **Python:** 3.6.9
* **Platform:** linux-ubuntu18.04-zen2
* **Concretizer:** clingo
```

**Comment by healther** [comment]
I'm not sure if this should be part of this discussion or its own: We ran into [GTKPlus](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/gtkplus/package.py#L9) now being a MesonPackage, which _requires_ py3, but we need it in (effectively) environments with py2. We "solved" this by just retaining a pre-MesonPackage version of gtkplus/package.py

I think this kind of problem would also be solved with your solution here. But it might be worth thinking about relaxing the "A package has a single buildtype" assumption behind our BasePackage class.
